gpcontrib/gp_orca/gporca.c

```diff
@@ -112,6 +137,113 @@ gp_orca_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 	return result;
 }
 
+/*
+ * gp_orca_explain_dxl -
+ *	  print out the execution plan for one Query in DXL format
+ *	  this function implicitly uses optimizer
+ */
+static void
+gp_orca_explain_dxl(Query *query, ExplainState *es, const char *queryString,
+					ParamListInfo params)
+{
+	MemoryContext oldcxt = CurrentMemoryContext;
+	bool save_enumerate;
+	char *dxl = NULL;
+	PlannerInfo *root;
+	PlannerGlobal *glob;
+	Query *pqueryCopy;
+
+	save_enumerate = optimizer_enumerate_plans;
+
+	/* Do the EXPLAIN. */
+
+	/* enable plan enumeration before calling optimizer */
+	optimizer_enumerate_plans = true;
+
+	/*
+	 * Initialize a dummy PlannerGlobal struct. ORCA doesn't use it, but the
+	 * pre- and post-processing steps do.
+	 */
+	glob = makeNode(PlannerGlobal);
+	glob->subplans = NIL;
+	glob->subroots = NIL;
+	glob->rewindPlanIDs = NULL;
+	glob->transientPlan = false;
+	glob->oneoffPlan = false;
+	glob->share.shared_inputs = NULL;
+	glob->share.shared_input_count = 0;
+	glob->share.motStack = NIL;
+	glob->share.qdShares = NULL;
+	/* these will be filled in below, in the pre- and post-processing steps */
+	glob->finalrtable = NIL;
+	glob->relationOids = NIL;
+	glob->invalItems = NIL;
+
+	root = makeNode(PlannerInfo);
+	root->parse = query;
+	root->glob = glob;
+	root->query_level = 1;
+	root->planner_cxt = CurrentMemoryContext;
+	root->wt_param_id = -1;
+
+	/* create a local copy to hand to the optimizer */
+	pqueryCopy = (Query *) copyObject(query);
+
+	/*
+	 * Pre-process the Query tree before calling optimizer.
+	 *
+	 * Constant folding will add dependencies to functions or relations in
+	 * glob->invalItems, for any functions that are inlined or eliminated
+	 * away. (We will find dependencies to other objects later, after planning).
+	 */
+	pqueryCopy = fold_constants(root, pqueryCopy, params,
+								GPOPT_MAX_FOLDED_CONSTANT_SIZE);
+
+	/*
+	 * If any Query in the tree mixes window functions and aggregates, we need to
+	 * transform it such that the grouped query appears as a subquery
+	 */
+	pqueryCopy = (Query *) transformGroupedWindows((Node *) pqueryCopy, NULL);
+
+
+	/* optimize query using optimizer and get generated plan in DXL format */
+	dxl = SerializeDXLPlan(pqueryCopy);
+
+	/* restore old value of enumerate plans GUC */
+	optimizer_enumerate_plans = save_enumerate;
+
+	if (dxl == NULL)
+		elog(NOTICE, "Optimizer failed to produce plan");
+	else
+	{
+		appendStringInfoString(es->str, dxl);
+		appendStringInfoChar(es->str, '\n'); /* separator line */
+		pfree(dxl);
+	}
+
+	/* Free the memory we used. */
+	MemoryContextSwitchTo(oldcxt);
+}
+
+static void
+gp_orca_explain(Query *query, int cursorOptions, IntoClause *into,
+				ExplainState *es, const char *queryString, ParamListInfo params,
+				QueryEnvironment *queryEnv)
+{
+	if (es->dxl)
```

Comment:
Perhaps:
```suggestion
	if (NULL != es->dxl)
```
And in also in similar places below.

Reply:
It is bool:
```
	bool		dxl;			/* CDB: print DXL */
```